tensorflow/contrib/azure/az_blob_file_system.cc

```diff
@@ -0,0 +1,701 @@
+    *container = objectp.ToString();
+    *object = "";
+  } else {
+    *container = objectp.substr(0, pos).ToString();
+    *object = objectp.substr(pos + 1).ToString();
+  }
+
+  return Status::OK();
+}
+
+std::string errno_to_string() {
+  switch (errno) {
+    case invalid_parameters:
+      return "invalid_parameters";
+    /* client level*/
+    case client_init_fail:
+      return "client_init_fail";
+    case client_already_init:
+      return "client_already_init";
+    case client_not_init:
+      return "client_not_init";
+    /* container level*/
+    case container_already_exists:
+      return "container_already_exists";
+    case container_not_exists:
+      return "container_not_exists";
+    case container_name_invalid:
+      return "container_name_invalid";
+    case container_create_fail:
+      return "container_create_fail";
+    case container_delete_fail:
+      return "container_delete_fail";
+    /* blob level*/
+    case blob__already_exists:
+      return "blob__already_exists";
+    case blob_not_exists:
+      return "blob_not_exists";
+    case blob_name_invalid:
+      return "blob_name_invalid";
+    case blob_delete_fail:
+      return "blob_delete_fail";
+    case blob_list_fail:
+      return "blob_list_fail";
+    case blob_copy_fail:
+      return "blob_copy_fail";
+    case blob_no_content_range:
+      return "blob_no_content_range";
+    /* unknown error*/
+    case unknown_error:
+    default:
+      return "unknown_error";
+  }
+}
+
+microsoft_azure::storage::blob_client_wrapper CreateAzBlobClientWrapper(
+    const std::string& account) {
+  const auto key = std::getenv("TF_AZURE_STORAGE_KEY");
+  std::shared_ptr<microsoft_azure::storage::storage_credential> credentials;
+  if (key != nullptr) {
+    credentials =
+        std::make_shared<microsoft_azure::storage::shared_key_credential>(
+            account, key);
+  } else {
+    credentials =
+        std::make_shared<microsoft_azure::storage::anonymous_credential>();
+  }
+  auto storage_account =
+      std::make_shared<microsoft_azure::storage::storage_account>(account,
+                                                                  credentials);
+  auto blob_client = std::make_shared<microsoft_azure::storage::blob_client>(
+      storage_account, 10);
+  microsoft_azure::storage::blob_client_wrapper blob_client_wrapper(
+      blob_client);
+
+  return blob_client_wrapper;
+}
+
+class AzBlobRandomAccessFile : public RandomAccessFile {
+ public:
+  AzBlobRandomAccessFile(const std::string& account, const std::string& container,
+                      const std::string& object)
+      : account_(account), container_(container), object_(object) {}
+
+  Status Read(uint64 offset, size_t n, StringPiece* result,
+              char* scratch) const override {
+    auto blob_client = CreateAzBlobClientWrapper(account_);
+    std::stringstream ss;
+    blob_client.download_blob_to_stream(container_, object_, offset, n, ss);
```

Comment:
@surry pushed a change now having
```c++
std::ostringstream oss;
oss.rdbuf()->pubsetbuf(scratch, n);
```
This is new to me, but does seem better than what I had before and profiling has less allocations than before. Let me know if you know of a better way than this.

Tests on Linux still pass, but running on macOS now fails as `scratch` is empty. Would you know why this might happen with this change?

Reply:
That's what I had in mind. I'm not sure why you'd only see a failure on macOS - maybe try a `std::ostream` instead of an ostringstream?

Reply:
All the `std::ostream` constructors are protected that we'd need here